moshi-adapters-kotlinx-immutable/src/main/kotlin/dev/zacsweers/moshix/immutable/adapters/ImmutableCollectionJsonAdapter.kt

```diff
@@ -0,0 +1,158 @@
+package dev.zacsweers.moshix.immutable.adapters
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonReader
+import com.squareup.moshi.JsonWriter
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import com.squareup.moshi.rawType
+import kotlinx.collections.immutable.ImmutableCollection
+import kotlinx.collections.immutable.ImmutableList
+import kotlinx.collections.immutable.ImmutableMap
+import kotlinx.collections.immutable.ImmutableSet
+import kotlinx.collections.immutable.PersistentCollection
+import kotlinx.collections.immutable.PersistentList
+import kotlinx.collections.immutable.PersistentMap
+import kotlinx.collections.immutable.PersistentSet
+import kotlinx.collections.immutable.persistentListOf
+import kotlinx.collections.immutable.persistentMapOf
+import kotlinx.collections.immutable.persistentSetOf
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+
+/**
+ * Applying [ImmutableCollectionJsonAdapterFactory] to your Moshi instance will allow serialization between
+ * JSON lists/map and these types:
+ *  - [kotlinx.collections.immutable.PersistentCollection]
+ *  - [kotlinx.collections.immutable.PersistentList]
+ *  - [kotlinx.collections.immutable.PersistentMap]
+ *  - [kotlinx.collections.immutable.PersistentSet]
+ *  - [kotlinx.collections.immutable.ImmutableCollection]
+ *  - [kotlinx.collections.immutable.ImmutableList]
+ *  - [kotlinx.collections.immutable.ImmutableMap]
+ *  - [kotlinx.collections.immutable.ImmutableSet]
+ *
+ * Example:
+ * ```
+ * val moshi = Moshi.Builder()
+ *   .add(PersistentCollectionJsonAdapterFactory)
+ *   .build()
+ *
+ * @JsonClass(generateAdapter = true)
+ * data class Data(
+ *   val stringList: ImmutableList<String>, // or PersistentList<String>,
+ *   val objList: ImmutableList<SomeObject>, // or PersistentList<SomeObject>,
+ *   val objCollection: ImmutableCollection<SomeObject>, // or PersistentCollection<SomeObject>,
+ *   val objMap: ImmutableMap<String, SomeObject>, // or PersistentMap<String, SomeObject>,
+ *   val objectsSet: ImmutableMap<SomeObject>, // or PersistentMap<SomeObject>,
+ * )
+ * ```
+ */
+
+public class ImmutableCollectionJsonAdapterFactory : JsonAdapter.Factory {
```

Comment:
This should be called ImmutableCollectionsJsonAdapterFactory (note the plural name)